### PR TITLE
MethodSeparationFixer - add class elements separation options

### PR DESCRIPTION
**Pull request by SpacePossum**
This PR adds options to the `MethodSeparationFixer`. New options are; `const` and `property` and `method` is the default.
The separation of this elements is the same as the methods are currently done with one exception;

```php
class Sample
{
    private $a; // this comment is allowed here

    private function doSomething() {
         //
    }  // this comment is still not allowed and moved down
}

```

Minor change is the fixer now allows comments after a method in an interface
Sample:

```php
interface SomeThing
{
    public function doSomething(); // this comment is now allowed
}
```

* There might be a BC break in the TokensAnalyzer, not sure.
* ~We might want to rename the MethodSeparationFixer in 3.0.~

closes https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/2219 (cc @localheinz)

### CODE REVIEW

File: src/Fixer/ClassNotation/MethodSeparationFixer.php [hunk 1]
```diff
@@ -76,159 +113,187 @@ public function getPriority()
...
-        }
+        $types = ['const', 'method', 'property'];
+
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder('elements', sprintf('List of classy elements; \'%s\'.', implode("', '", $types))))
+                ->setAllowedTypes(['array'])
+                ->setAllowedValues([
+                    (new FixerOptionValidatorGenerator())->allowedValueIsSubsetOf($types),
+                ])
+                ->setDefault(['method'])
```
Comment (keradus) [inline src/Fixer/ClassNotation/MethodSeparationFixer.php]: please add `TODO` for:
> We might want to rename the MethodSeparationFixer in 3.0.

and to switch default behaviour to fix all types

**Comment by SpacePossum** [comment]
`fabbot.io` is false negative because we updated the CS rules of SF set but this not yet part of `fabbot.io`

File: README.rst [hunk 2]
```diff
@@ -645,7 +645,13 @@ Choose from the list of available rules:
 
 * **method_separation** [@Symfony]
 
-  Methods must be separated with one blank line.
```
Comment (keradus) [inline README.rst]: we should rename the fixer

File: src/Fixer/ClassNotation/MethodSeparationFixer.php [hunk 3]
```diff
@@ -22,20 +26,30 @@
...
  *
  * @author SpacePossum
  */
-final class MethodSeparationFixer extends AbstractFixer implements WhitespacesAwareFixerInterface
+final class MethodSeparationFixer extends AbstractFixer implements ConfigurationDefinitionFixerInterface, WhitespacesAwareFixerInterface
 {
     /**
+     * @var array<string, true>
+     */
+    private $classElementTypes = [];
```
Comment (keradus) [inline src/Fixer/ClassNotation/MethodSeparationFixer.php]: do we need this property? can't we rely just on `$this->configuration['elements']` directly ?

File: src/Fixer/ClassNotation/MethodSeparationFixer.php [hunk 4]
```diff
@@ -76,159 +113,187 @@ public function getPriority()
...
-        }
+        $types = ['const', 'method', 'property'];
+
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder('elements', sprintf('List of classy elements; \'%s\'.', implode("', '", $types))))
+                ->setAllowedTypes(['array'])
+                ->setAllowedValues([
+                    (new FixerOptionValidatorGenerator())->allowedValueIsSubsetOf($types),
+                ])
+                ->setDefault(['method']) // TODO Rename the MethodSeparationFixer in 3.0. Set default to ['const', 'method', 'property']
```
Comment (keradus) [inline src/Fixer/ClassNotation/MethodSeparationFixer.php]: I would make it already on 2.x line to decrease amount of work needed to migrate 2.x->3.x
that way, on 2.x we will have future-compat layer already.

rename the fixer already and then create `MethodSeparationFixer` once more that will be only a proxy fixer with unmodified behaviour, no configuration options, and will be deprecated.

File: src/Tokenizer/TokensAnalyzer.php [hunk 5]
```diff
@@ -535,8 +535,8 @@ private function findClassyElements(Tokens $tokens, $start)
                 continue;
             }
 
-            if (!$inClass) {
-                $inClass = $token->isClassy();
+            if (false === $inClass) {
+                $inClass = $token->isClassy() ? $index : false;
```
Comment (keradus) [inline src/Tokenizer/TokensAnalyzer.php]: concern (a little out of the blue):
does this method supports anonymous classes?

File: src/Tokenizer/TokensAnalyzer.php [hunk 6]
```diff
@@ -580,15 +580,27 @@ private function findClassyElements(Tokens $tokens, $start)
             }
 
             if (0 === $bracesLevel && $token->isGivenKind(T_VARIABLE)) {
-                $elements[$index] = ['token' => $token, 'type' => 'property'];
+                $elements[$index] = [
+                    'token' => $token,
+                    'type' => 'property',
+                    'class' => $inClass,
```
Comment (keradus) [inline src/Tokenizer/TokensAnalyzer.php]: what is here? boolean ?

File: src/Fixer/ClassNotation/MethodSeparationFixer.php [hunk 7]
```diff
@@ -76,159 +113,187 @@ public function getPriority()
...
+        return $tokens->isAnyTokenKindsFound(Token::getClassyTokenKinds());
+    }
+
+    /**
+     * {@inheritdoc}
+     */
     protected function applyFix(SplFileInfo $file, Tokens $tokens)
     {
         $tokensAnalyzer = new TokensAnalyzer($tokens);
+        $class = $classStart = $classEnd = false;
```
Comment (keradus) [inline src/Fixer/ClassNotation/MethodSeparationFixer.php]: what is `$class` here if not start nor end ?

Comment (SpacePossum) [reply src/Fixer/ClassNotation/MethodSeparationFixer.php]: `$class` value doesn't matter here, it is to keep SCA happy. Whenever `$class` is used as value it is the index of a classy (`T_CLASS`, `T_INTERFACE`, `T_TRAIT`). It cannot be otherwise unless invalid Tokens collection is used (i.e. tokenized invalid PHP).

File: src/Tokenizer/TokensAnalyzer.php (hunk 6 above)
Comment (SpacePossum) [reply src/Tokenizer/TokensAnalyzer.php]: `int`, index in the Tokens collection of the `classy` token

File: src/Tokenizer/TokensAnalyzer.php (hunk 5 above)
Comment (SpacePossum) [reply src/Tokenizer/TokensAnalyzer.php]: yes it does, anonymous classes use the same `T_CLASS` token as regular classes;
```bash
$ php test.php 

    0 | T_OPEN_TAG                              | <?php

    1 | T_WHITESPACE                            | 

    2 | T_VARIABLE                              | $a
    3 | T_WHITESPACE                            |  
    4 |                                         | =
    5 | T_WHITESPACE                            |  
    6 | T_NEW                                   | new
    7 | T_WHITESPACE                            |  
    8 | T_CLASS                                 | class
    9 | T_WHITESPACE                            |  
   10 |                                         | {
   11 | T_PUBLIC                                | public
```

File: src/Fixer/ClassNotation/MethodSeparationFixer.php (hunk 7 above)
Comment (keradus) [reply src/Fixer/ClassNotation/MethodSeparationFixer.php]: then what's the difference between `$class` and `$classStart` ?

File: src/Tokenizer/TokensAnalyzer.php (hunk 5 above)
Comment (keradus) [reply src/Tokenizer/TokensAnalyzer.php]: it's not question in terms of `isClassy` implementation details, but fixer implementation and coverage by tests.

File: src/Tokenizer/TokensAnalyzer.php (hunk 6 above)
Comment (keradus) [reply src/Tokenizer/TokensAnalyzer.php]: please add `index` or sth...

File: src/Fixer/ClassNotation/MethodSeparationFixer.php (hunk 7 above)
Comment (SpacePossum) [reply src/Fixer/ClassNotation/MethodSeparationFixer.php]: >then what's the difference between $class and $classStart ?

`$class` is provided by the analyzer, `$classStart` is the index of the opening `{` of the class.

File: src/Tokenizer/TokensAnalyzer.php (hunk 5 above)
Comment (SpacePossum) [reply src/Tokenizer/TokensAnalyzer.php]: this is the analyzer code, so I don't understand, it is covered by the analyzer tests

File: src/Fixer/ClassNotation/MethodSeparationFixer.php (hunk 1 above)
Comment (SpacePossum) [reply src/Fixer/ClassNotation/MethodSeparationFixer.php]: done and removed after additional request

File: README.rst (hunk 2 above)
Comment (SpacePossum) [reply README.rst]: done

File: src/Fixer/ClassNotation/MethodSeparationFixer.php (hunk 3 above)
Comment (SpacePossum) [reply src/Fixer/ClassNotation/MethodSeparationFixer.php]:
```
array(3) {
  [0]=>
  string(5) "const"
  [1]=>
  string(6) "method"
  [2]=>
  string(8) "property"
}
```

```
array(3) {
  ["const"]=>
  bool(true)
  ["method"]=>
  bool(true)
  ["property"]=>
  bool(true)
}
```

i.e. direct usage would be `in_array($type, $this->config['elements'], true)` vs. `isset($classElementTypes[$type])`, the last one is faster

File: src/Tokenizer/TokensAnalyzer.php (hunk 6 above)
Comment (SpacePossum) [reply src/Tokenizer/TokensAnalyzer.php]: done

File: src/Tokenizer/TokensAnalyzer.php [hunk 8]
```diff
@@ -604,6 +604,8 @@ private function findClassyElements(Tokens $tokens, $start)
             }
         }
 
+        ksort($elements);
```
Comment (SpacePossum) [inline src/Tokenizer/TokensAnalyzer.php]: TODO: send this bug fix to 2.2 line as well

File: src/Fixer/ClassNotation/ClassAttributesSeparationFixer.php [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function createConfigurationDefinition()
+    {
+        $types = ['const', 'method', 'property'];
+
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder('elements', sprintf('List of classy elements; \'%s\'.', implode("', '", $types))))
```
Comment (julienfalque) [inline src/Fixer/ClassNotation/ClassAttributesSeparationFixer.php]: I'm not a big fan of this option, it's not flexible enough IMO. I would replace it with something like:
```php
[
    'const' => 'no_blank_line | single_blank_line | single_blank_line_if_phpdoc',
    'property' => 'no_blank_line | single_blank_line | single_blank_line_if_phpdoc',
    'method' => 'no_blank_line | single_blank_line | single_blank_line_if_phpdoc',
]
```
WDYT?

File: tests/Fixtures/Integration/priority/ordered_class_elements,class_attributes_separation.test [hunk 10]
```diff
@@ -1,7 +1,7 @@
 --TEST--
-Integration of fixers: ordered_class_elements,method_separation.
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply tests/Fixtures/Integration/priority/ordered_class_elements,class_attributes_separation.test]: do we keep those for the other deprecated fixers? seems like a lot of copy-paste for not much value to me...

File: src/Fixer/ClassNotation/ClassAttributesSeparationFixer.php (hunk 9 above)
Comment (SpacePossum) [reply src/Fixer/ClassNotation/ClassAttributesSeparationFixer.php]: I prefer to add one feature in one PR, here allow not _only_ methods but also other class elements to be separated by a single line. When this is done I'm happy to think about more options if there is a demand from a community for it ;)

File: tests/Fixtures/Integration/priority/ordered_class_elements,class_attributes_separation.test (hunk 10 above)
[the remark this answers is not in the document]
Comment (julienfalque) [reply tests/Fixtures/Integration/priority/ordered_class_elements,class_attributes_separation.test]: This has value: deprecating a fixer does not mean we allow it to break. If we removed tests for other deprecated fixers, that's a mistake IMO.

[the remark this answers is not in the document]
Comment (SpacePossum) [reply tests/Fixtures/Integration/priority/ordered_class_elements,class_attributes_separation.test]: Having the fixture test `A+B` and `C+B` where `A === C` has little value, as we know the `A === C` because we test this as `A` is now proxied to `C`, no?

[the remark this answers is not in the document]
Comment (julienfalque) [reply tests/Fixtures/Integration/priority/ordered_class_elements,class_attributes_separation.test]: `A` being proxied to `C` is an implementation detail. The test is here to proove `A` continues to work as expected, no matter how it is implemented. Testing `C` does not mean `A` uses it correctly.

File: tests/AutoReview/FixerFactoryTest.php [hunk 11]
```diff
@@ -58,8 +58,8 @@ public function provideFixersPriorityCases()
 
         $cases = [
             [$fixers['elseif'], $fixers['braces']],
-            [$fixers['method_separation'], $fixers['braces']],
-            [$fixers['method_separation'], $fixers['indentation_type']],
```
[the remark this answers is not in the document]
Comment (keradus) [reply tests/AutoReview/FixerFactoryTest.php]: we should

File: tests/Fixtures/Integration/priority/ordered_class_elements,class_attributes_separation.test (hunk 10 above)
[the remark this answers is not in the document]
Comment (keradus) [reply tests/Fixtures/Integration/priority/ordered_class_elements,class_attributes_separation.test]: I agree with @julienfalque here
it's extra easy to drop tests for deprecated fixer when it got removed. we keep tests for other deprecated fixers. 
imagine that in future one will change ProxyFixer implementation, or A is proxied to B with configuration. we need to know it works the same as it was, so we change implementation, but keep tests.

[the remark this answers is not in the document]
Comment (SpacePossum) [reply tests/Fixtures/Integration/priority/ordered_class_elements,class_attributes_separation.test]: so IIUC, I should restore the old fixtures-files and than copy-paste those using the new fixer name and configuration so we've both?

[the remark this answers is not in the document]
Comment (julienfalque) [reply tests/Fixtures/Integration/priority/ordered_class_elements,class_attributes_separation.test]: Yes.

[the remark this answers is not in the document]
Comment (SpacePossum) [reply tests/Fixtures/Integration/priority/ordered_class_elements,class_attributes_separation.test]: and how to deal with deprecation notices raised in the integration tests than?

[the remark this answers is not in the document]
Comment (julienfalque) [reply tests/Fixtures/Integration/priority/ordered_class_elements,class_attributes_separation.test]: Tests with deprecated fixers should be moved to a specific method in the test class with a `@group legacy` annotation. But I think fixers don't trigger deprecation notices so you might not have to change anything. Though we should take care of it in a separate PR.

File: src/Tokenizer/TokensAnalyzer.php (hunk 5 above)
Comment (SpacePossum) [reply src/Tokenizer/TokensAnalyzer.php]: > fixer implementation and coverage by tests.

tests added for `anonymous classes` for the fixer implementation

**Comment by SpacePossum** [comment]
interesting 7.2 failure https://travis-ci.org/FriendsOfPHP/PHP-CS-Fixer/jobs/289951049#L697

**Comment by keradus** [comment]
too many deprecations perhaps ?

**Comment by SpacePossum** [comment]
`phpdbg` with `phpunit` coverage runs out of memory, this is a known issue.

**Comment by keradus** [comment]
obviously phpunit with phpdbg/xdebug demands more memory, yet on master we don't demand that much it crashes the build

**Comment by SpacePossum** [comment]
https://github.com/sebastianbergmann/phpunit/issues/2019
and
https://stackoverflow.com/questions/46001411/phpdbg-phpunit-throws-error-even-when-memory-limit-is-1

If I run the `phpdbg phpunit` setup local it crashes as well, running with `xdebug phpunit` setup there is no issue. Not sure how to proceed for now.

**Comment by keradus** [comment]
if we don't get memory issue with phpdbg on master branch and we do on this PR, then I would say this PR requires too much memory. could we check it and shrink it down perhaps ?

**Comment by SpacePossum** [comment]
I don't think the issue is within this PR, I think this PR just pushes the total amount of memory used over the total limit.

File: src/Tokenizer/TokensAnalyzer.php (hunk 8 above)
Comment (keradus) [reply src/Tokenizer/TokensAnalyzer.php]: what is the bug actually ?

**Comment by keradus** [comment]
about memory issue.
memory needed by PHPUnit on master: 108MB, on this branch: 108MB

File: src/Tokenizer/TokensAnalyzer.php (hunk 8 above)
Comment (SpacePossum) [reply src/Tokenizer/TokensAnalyzer.php]: When there are anonymous classes within a class which also has methods before (or after) the anonymous class, the returned elements are not sorted.
However most fixers that use this method assume the returned `[index, [...]]` array is sorted on the index; low-to-high. When out of order some edge cases will throw the fixers off when they iterate over the returned value and insert/remove tokens.

File: tests/Fixer/ClassNotation/ClassAttributesSeparationFixerTest.php [hunk 12]
```diff
@@ -957,24 +957,44 @@ class A
...
         $this->doTest($expected, $input);
     }
 
-    public function provideFix70Cases()
+    public function provideFix70BigCases()
     {
         $to = $from = '<?php ';
+
         for ($i = 0; $i < 15; ++$i) {
             $from .= sprintf('class A%d{public function AA%d(){return new class {public function BB%d(){}};}public function otherFunction%d(){}}', $i, $i, $i, $i);
```
Comment (SpacePossum) [inline tests/Fixer/ClassNotation/ClassAttributesSeparationFixerTest.php]: note to self: `%1$d` should be used and not multiple `$i` as args.

File: tests/Fixer/ClassNotation/ClassAttributesSeparationFixerTest.php [hunk 13]
```diff
@@ -0,0 +1,1094 @@
...
+    {
+        $this->doTest($expected, $input);
+    }
+
+    public function provideFix70BigCases()
+    {
+        $to = $from = '<?php ';
+
+        for ($i = 0; $i < 15; ++$i) {
+            $from .= sprintf('class A%d{public function AA%d(){return new class {public function BB%d(){}};}public function otherFunction%d(){}}', $i, $i, $i, $i);
```
Comment (SpacePossum) [inline tests/Fixer/ClassNotation/ClassAttributesSeparationFixerTest.php]: note to self: %1$d should be used and not multiple $i as args.
but more important, this blows up the memory usage of the fixer to extreme values

Comment (keradus) [reply tests/Fixer/ClassNotation/ClassAttributesSeparationFixerTest.php]: actually, string interpolation is better than sprintf on php 7

Comment (keradus) [reply tests/Fixer/ClassNotation/ClassAttributesSeparationFixerTest.php]: 3199 is supposed to fix it? if so, please cherry-pick it to this PR and revert my "big test" so we will see would it work indeed ;)

**Comment by keradus** [comment]
Thank you @SpacePossum.

**Comment by TomasVotruba** [comment]
Great job @SpacePossum :heart:  , finally can drop my property and constant fixer

**Comment by SpacePossum** [comment]
Thanks for the ping back @TomasVotruba , always makes me happy to hear people like something I PR'ed :)  :+1:

**Comment by keradus** [comment]
hehe, hint for next time @TomasVotruba : help with a review, especially when you already played with familiar concept, and feature would be finished sooner ;)

**Comment by TomasVotruba** [comment]
@keradus I don't follow this repo, but feel free to ping me if you'd like my review ;)

**Comment by keradus** [comment]
so dead PR ;P
my comment was connected to yours
> finally can drop my property and constant fixer

we have no clue what are those, if you have your custom fixer that you don't want to contribute, yet want to have it done by someone else to main repo, review is helpful.

**Comment by TomasVotruba** [comment]
Sure, just ping me if need my review
